apps/server/src/modules/files-storage/service/preview.service.ts

```diff
@@ -0,0 +1,110 @@
+import { Injectable, InternalServerErrorException, NotAcceptableException } from '@nestjs/common';
+import { LegacyLogger } from '@src/core/logger';
+import crypto from 'crypto';
+import { subClass } from 'gm';
+import { PassThrough } from 'stream';
+import { S3ClientAdapter } from '../client/s3-client.adapter';
+import { DownloadFileParams, PreviewParams } from '../controller/dto';
+import { FileRecord } from '../entity';
+import { ErrorType } from '../error';
+import { IGetFile, IGetFileResponse } from '../interface';
+import { PreviewOutputMimeTypes } from '../interface/preview-output-mime-types.enum';
+import { FileDtoBuilder, FileResponseBuilder } from '../mapper';
+import { FilesStorageService } from './files-storage.service';
+
+@Injectable()
+export class PreviewService {
+	constructor(
+		private readonly storageClient: S3ClientAdapter,
+		private readonly fileStorageService: FilesStorageService,
+		private logger: LegacyLogger
+	) {
+		this.logger.setContext(PreviewService.name);
+	}
+
+	private checkIfPreviewPossible(fileRecord: FileRecord): void | NotAcceptableException {
+		if (!fileRecord.isPreviewPossible()) {
+			this.logger.warn(`could not generate preview for : ${fileRecord.id} ${fileRecord.mimeType}`);
+			throw new NotAcceptableException(ErrorType.PREVIEW_NOT_POSSIBLE);
+		}
+	}
+
+	public async getPreview(
+		fileRecord: FileRecord,
+		params: DownloadFileParams,
+		previewParams: PreviewParams,
+		bytesRange?: string
+	): Promise<IGetFileResponse> {
+		this.checkIfPreviewPossible(fileRecord);
+
+		const hash = this.createNameHash(params, previewParams);
+		const filePath = ['previews', fileRecord.getSchoolId(), hash].join('/');
+		const fileNameWithoutExtension = fileRecord.name.split('.')[0];
+		const format = this.getFormat(previewParams.outputFormat);
+		const name = `${fileNameWithoutExtension}.${format}`;
+		let file: IGetFile;
+
+		try {
+			file = await this.storageClient.get(filePath, bytesRange);
+		} catch (error) {
+			this.throwIfOtherError(error);
+
+			file = await this.generatePreview(fileRecord, params, previewParams, hash, filePath, bytesRange);
+		}
+
+		const response = FileResponseBuilder.build(file, name);
+
+		return response;
+	}
+
+	private throwIfOtherError(error): void | InternalServerErrorException {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+		if (error.message && error.message !== 'NoSuchKey') {
```

Comment:
This condition needs to be
```suggestion
		if (!error.message || error.message !== 'NoSuchKey') {
```
..if you want to keep it. Like stated above I suggest to check elsewhise.